modules/dag-l0/src/main/scala/org/tessellation/dag/l0/domain/snapshot/programs/GlobalSnapshotEventCutter.scala

```diff
@@ -0,0 +1,113 @@
+import cats.syntax.traverse._
+import cats.syntax.unorderedFoldable._
+
+import org.tessellation.dag.l0.infrastructure.snapshot.{DAGEvent, StateChannelEvent}
+import org.tessellation.ext.kryo._
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.node.shared.domain.event.EventCutter
+import org.tessellation.schema.Block
+import org.tessellation.security.hash._
+import org.tessellation.security.{Hashed, Hasher}
+
+import eu.timepit.refined.types.numeric.PosInt
+
+object GlobalSnapshotEventCutter {
+  def make[F[_]: Async: KryoSerializer: Hasher](maxBinarySizeBytes: PosInt): EventCutter[F, StateChannelEvent, DAGEvent] =
+    new EventCutter[F, StateChannelEvent, DAGEvent] {
+
+      def cut(scEvents: List[StateChannelEvent], dagEvents: List[DAGEvent]): F[(List[StateChannelEvent], List[DAGEvent])] =
+        doesNotExceedMaxSize(scEvents, dagEvents)
+          .ifM((scEvents, dagEvents).pure[F], removeEvents(scEvents, dagEvents))
+
+      def doesNotExceedMaxSize(scEvents: List[StateChannelEvent], dagEvents: List[DAGEvent]): F[Boolean] =
+        (scEvents.toBinary.liftTo[F], dagEvents.toBinary.liftTo[F])
+          .mapN((a, b) => a.length + b.length <= maxBinarySizeBytes.value)
+
+      def removeEvents(scEvents: List[StateChannelEvent], dagEvents: List[DAGEvent]): F[(List[StateChannelEvent], List[DAGEvent])] =
+        (
+          scEvents.traverse(evt => evt.pure[F].product(StateChannelEventWithFee(evt))).map(_.toMap),
+          dagEvents.traverse(evt => evt.pure[F].product(DAGBlockEventWithFee(evt))).map(_.toMap)
+        ).flatMapN {
+          case (scEventsMap, dagEventsMap) =>
+            def removeOne(scEvents: List[StateChannelEvent], dagEvents: List[DAGEvent]): (List[StateChannelEvent], List[DAGEvent]) =
+              (scEvents, dagEvents) match {
+                case (Nil, Nil)    => (Nil, Nil)
+                case (_ :: t, Nil) => (t, Nil)
+                case (Nil, _ :: t) => (Nil, t)
+                case (h1 :: t1, h2 :: t2) =>
+                  Ordering[EventWithFee].compare(scEventsMap(h1), dagEventsMap(h2)) match {
+                    case n if (n == -1) || (n == 0 && t1.length >= t2.length) => (t1, dagEvents)
+                    case _                                                    => (scEvents, t2)
+                  }
+              }
+
+            val scEventsSorted = scEvents.sortWith((left, right) => Ordering[EventWithFee].lt(scEventsMap(left), scEventsMap(right)))
+            val dagEventsSorted = dagEvents.sortWith((left, right) => Ordering[EventWithFee].lt(dagEventsMap(left), dagEventsMap(right)))
+
+            (scEventsSorted, dagEventsSorted).tailRecM {
+              case (Nil, Nil) => MonadThrow[F].raiseError(new IllegalStateException("No events to cut"))
+
+              case (lefts, rights) =>
+                val result = removeOne(lefts, rights)
+                doesNotExceedMaxSize(result._1, result._2).map(flag => Either.cond(flag, result, result))
```

Comment:
Why is the outcome of `doesNotExceedMaxSize` called `flag`? Maybe just write it as `.map(Either.cond(_, result, result)` and don't introduce any naming.

Reply:
done